src/pages/api/token.ts

```diff
@@ -0,0 +1,85 @@
+import jwt from "jsonwebtoken";
+import { NextApiResponse } from "next";
+import { Handler } from "next-iron-session";
+import { ExtendedRequest } from "src/@types/ExtendedRequest";
+import withSession from "src/middleware/withSession";
+
+function promisifiedJWTVerify(
+  oauthData: { clientSecret?: string; accessToken: string },
+  request: ExtendedRequest,
+  response: NextApiResponse,
+): Promise<any> {
+  const { clientSecret, accessToken } = oauthData;
+
+  return new Promise((resolve, reject) => {
+    if (clientSecret) {
+      jwt.verify(
+        accessToken,
+        clientSecret,
+        async (error: jwt.VerifyErrors | null, decoded?: any) => {
+          if (error !== null) {
+            throw new Error(error.message);
+          }
+
+          request.session.set("user-id", decoded.sub);
+          await request.session.save().catch((error) => {
+            throw new Error(error);
+          });
+          response.writeHead(302, { Location: "/account" });
+          response.end();
+
+          resolve();
```

Comment:
I would prefer to have the function only do the reading of the JWT and have this session part be on the handler.

you could replace those lines with
```
resolve(decoded)
```

This would allow you to make the `promisifiedJWTVerify` function only take the JWT and the clientSecret.
What do you think?

Reply:
You are absolutely right, as my function name implies, it's just there to promisify `jwt.verif`, not handle the session.

Reply:
2aa9c0e

Feels weird to add `async` in the promise response, but I think I have no choice.